### ISSUE TITLE
Question: Docker SDK image for targeting both netstandard1.3 and net451?

### ISSUE THREAD

**Issue by eric-davis** [Microsoft/dotnet#432]
I am trying to upgrade some of my existing projectjson dotnet core library projects to VS2017 tooling and the new csproj format. Because of some of the 3rd-party libraries I am using, some of these projects target both netstandard1.3 and net451. My CI build process involves automated builds in Docker containers using a Docker custom image that includes both the dotnet-1.1.0-sdk-projectjson image merged with a mono installation. This process has worked well up til now. Now that I am trying to upgrade, I created a new Docker build image that consists of the dotnet-1.1.2-sdk image with the same mono image - see Docker file below:

```
FROM microsoft/dotnet:1.1.2-sdk

RUN apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF \
	&& echo "deb http://download.mono-project.com/repo/debian wheezy-libjpeg62-compat main" | tee -a /etc/apt/sources.list.d/mono-xamarin.list \
	&& echo "deb http://download.mono-project.com/repo/debian wheezy/snapshots/4.8.1 main" | tee -a /etc/apt/sources.list.d/mono-xamarin.list \
	&& apt-get update \
	&& apt-get install -y msbuild mono-devel referenceassemblies-pcl ca-certificates-mono mono-xsp4 nuget \
	&& rm -rf /var/lib/apt/lists/*
```

Unfortunately, the builds are failing as a result of targeting net451 with the following error:

`error MSB3644: The reference assemblies for framework ".NETFramework,Version=v4.5.1" were not found. To resolve this, install the SDK or Targeting Pack for this framework version or retarget your application to a version of the framework for which you have the SDK or Targeting Pack installed. Note that assemblies will be resolved from the Global Assembly Cache (GAC) and will be used in place of reference assemblies. Therefore your assembly may not be correctly targeted for the framework you intend.` 

What am I missing? Or, better yet, are there any pre-built, official Docker SDK images out there that will meet my needs? I haven't been able to find any on my own to this point.

Thanks!

**Comment by svick** [comment]
I take it you're looking for a Linux image and a Windows image wouldn't help?

**Comment by richlander** [comment]
We don't currently have a good way of delivering .NET Framework reference assemblies for this purpose.  It is understood that this is necessary/important and we have been talking about how to solve it. 

I think that there are two main options (not mutually exclusive) for my team:

- Release a reference assemblies zip that can be copied and expanded into your own Docker images.
- Release a .NET build-specific Docker image that includes the reference assemblies.

We need to do the first to enable the second, so it's really just a question on whether the second option is valuable enough. We'll also guide folks down the multi-stage build path instead of creating a "works for all scenarios" Docker image. That will make your Docker files more complicated, but will avoid downloading a mega monster. Also, we'll need to decide on the supported scenarios for Linux vs. Windows.

Thoughts?

/cc @kendrahavens @terrajobst @MichaelSimons

**Comment by kendrahavens** [comment]
Hi @eric-davis, this issue has come up before for folks moving to the .NET CLI built on msbuild (csproj-based) who want to target .NET framework reference assemblies with mono. A few [work arounds](https://github.com/dotnet/sdk/issues/335#issuecomment-291487227) were worked out, but there appears to be a [CLI Issue](https://github.com/dotnet/cli/issues/5723) tracking this to make it more accessible though it doesn't appear to be high priority currently.

@richlander It would be excellent to wrap this in a Docker image to save folks the configuration time. If the different version combinations become numerous and a tagging issue we could at least post docs outlining the Dockerfiles that would cover the scenarios.

**Comment by eric-davis** [comment]
Thanks @richlander & @kendrahavens for your input. This is quickly becoming a high-priority issue for myself and my team. We would like to allow everyone to upgrade to VS2017 and the updated tooling, but can't do that until we can migrate our existing code and processes to support the csproj-based format. The suggestions regarding the creation of packaged Docker images or the walkthrough docs would be extremely helpful.

@kendrahavens - Thanks for the links to the workaround thread, but unfortunately that seems to be above my current level of understanding at the moment.

**Comment by kendrahavens** [comment]
Gotcha, the previous thread also has some votes asking for this so it's good to know multiple teams are held up.

**Comment by richlander** [comment]
Clarifying question ... is there anything special with the libraries you need to build? Most libraries just need the reference assemblies. Others might need tools if the libraries expose WCF services, include Xaml pages or rely on additional SDKs. We don't have any immediate plans on solving the extra scenarios within containers.

**Comment by eric-davis** [comment]
We require no special libraries to build at this time - just the reference assemblies.

**Comment by richlander** [comment]
@eric-davis  Did you try the workaround that @kendrahavens pointed to? If we can get you unblocked in the immediate term, that would be good, while we look at providing a better solution.

**Comment by eric-davis** [comment]
@richlander I did take a look at the workaround @kendrahavens pointed to. But honestly, it's a bit above my current level of understanding. For now, we have decided to just target netstandard1.3. We will just upgrade to .net 4.6 for any of our 4.51 projects that need to consume our new .net core library packages.

**Comment by richlander** [comment]
@eric-davis Fair enough. We'll continue to work on a better solution. I wrote up a new issue for it: https://github.com/dotnet/cli/issues/7023. 

I'll close this issue and we'll use dotnet/clr 7023 for tracking.

**Comment by eric-davis** [comment]
Thanks.
